test/docker-e2e/e2e_test.go

```diff
@@ -171,3 +187,256 @@ func (s *CelestiaTestSuite) WaitForSync(ctx context.Context, statusClient rpccli
 		}
 	}
 }
+
+// CheckLiveness validates that all validators proposed blocks and no nodes halted.
+// Automatically waits for sufficient blocks (3 per validator minimum) if needed.
+//
+// Upgrade-agnostic: can be called before/after upgrades or spanning the entire period.
+// Call at the end of E2E tests to validate network health.
+func (s *CelestiaTestSuite) CheckLiveness(
+	ctx context.Context,
+	chain tastoratypes.Chain,
+	startHeight int64,
+) error {
+	rpcClient, err := chain.GetNodes()[0].GetRPCClient()
+	if err != nil {
+		return fmt.Errorf("failed to get RPC client: %w", err)
+	}
+
+	endHeight, err := s.ensureMinimumBlocks(ctx, chain, rpcClient, startHeight)
+	if err != nil {
+		return fmt.Errorf("failed to ensure minimum blocks: %w", err)
+	}
+
+	startValidators, endValidators, err := s.fetchValidatorSets(ctx, rpcClient, startHeight, endHeight)
+	if err != nil {
+		return fmt.Errorf("failed to fetch validator sets: %w", err)
+	}
+
+	proposers, err := s.fetchProposerAddresses(ctx, rpcClient, startHeight, endHeight)
+	if err != nil {
+		return fmt.Errorf("failed to fetch proposer addresses: %w", err)
+	}
+
+	if err := s.validateAllValidatorsProposed(startValidators, endValidators, proposers, startHeight, endHeight); err != nil {
+		return err
+	}
+
+	if err := s.validateNodesNotHalted(ctx, chain, endHeight); err != nil {
+		return err
+	}
+
+	s.T().Logf("✅ Liveness check passed: all validators proposed blocks and all nodes reached height %d", endHeight)
```

Comment:
Please avoid using emojis in logs.

```suggestion
	s.T().Logf("Liveness check passed: all validators proposed blocks and all nodes reached height %d", endHeight)
```